presto-native-execution/presto_cpp/main/expression/RowExpressionEvaluator.cpp

```diff
@@ -0,0 +1,419 @@
+        isLeftNull = constantExpr->value()->isNullAt(0);
+        if (!isLeftNull) {
+          if (auto constVector =
+                  constantExpr->value()->as<ConstantVector<bool>>()) {
+            if (constVector->valueAt(0) == true) {
+              return getConstantExpression(constantExpr);
+            }
+            return right;
+          }
+        }
+      }
+
+      auto rexpr = compileExpression(right);
+      if (auto constantExpr =
+              std::dynamic_pointer_cast<const exec::ConstantExpr>(rexpr)) {
+        if (isLeftNull && constantExpr->value()->isNullAt(0)) {
+          return getConstantExpression(constantExpr);
+        }
+        if (auto constVector =
+                constantExpr->value()->as<ConstantVector<bool>>()) {
+          if (constVector->valueAt(0) == false) {
+            return left;
+          }
+          return right;
+        }
+      }
+      break;
+    }
+    case protocol::Form::IN:
+    case protocol::Form::DEREFERENCE:
+    case protocol::Form::SWITCH:
+    case protocol::Form::WHEN:
+    case protocol::Form::COALESCE:
+    case protocol::Form::ROW_CONSTRUCTOR:
+    case protocol::Form::BIND:
+      break;
+    default:
+      break;
+  }
+
+  return specialFormExpr;
+}
+
+json RowExpressionEvaluator::veloxExprToRowExpr(
+    const std::shared_ptr<const velox::exec::Expr>& expr,
+    const json& inputRowExpr) {
+  json res;
+  if (inputRowExpr.contains("sourceLocation")) {
+    res["sourceLocation"] = inputRowExpr["sourceLocation"];
+  }
+
+  if (expr->isConstant()) {
+    if (expr->type()->isRow()) {
+      // Velox constant expressions of ROW type map to special form expression
+      // row_constructor in Presto.
+      res = getSpecialFormRowConstructor(expr, inputRowExpr);
+    } else if (expr->inputs().empty()) {
+      res["@type"] = "constant";
+      auto constantExpr =
+          std::dynamic_pointer_cast<const exec::ConstantExpr>(expr);
+      VELOX_USER_CHECK_NOT_NULL(constantExpr);
+      auto type = constantExpr->type();
+      VectorPtr constVector = constantExpr->value();
+      res["valueBlock"] = serializeValueBlock(constVector);
+      res["type"] = expr->type()->toString();
+    } else {
+      // divide(0, 0) is a constant expression with constant inputs.
+      res = inputRowExpr;
+    }
+  } else if (
+      auto field =
+          std::dynamic_pointer_cast<const exec::FieldReference>(expr)) {
+    // variable
+    res = toVariableReferenceExpression(field.get());
+  } else if (expr->isSpecialForm()) {
+    // special
+    res = getSpecialForm(expr, inputRowExpr);
+  } else if (auto func = expr->vectorFunction()) {
+    // Check if special form expression or call expression.
+    std::unordered_set<std::string> specialForms = {
+        "in", "cast", "try", "is_null", "row_constructor"};
+    if (specialForms.count(expr->name()) != 0) {
+      // special
+      res = getSpecialForm(expr, inputRowExpr);
+    } else {
+      // call
+      res["@type"] = "call";
```

Comment:
Abstract this logic to a separate function. Would make this function limited to outer conversions only.